History bulk operations

davelopez opened the pull request:
Closes #13521

# Backend
 - New API route `PUT /api/histories/{history_id}/contents/bulk` that can apply an operation to multiple selected items in a History. The items to process can be explicitly set in the payload or selected by the filtering query parameters.

    ![Screenshot from 2022-03-18 16-28-33](https://user-images.githubusercontent.com/46503462/159032651-f7b417bc-3ed0-44c5-ac70-369a72d23f34.png)
 
 - The `GET /api/histories/{history_id}/contents` now accepts a custom media type (thanks to @mvdbeek) that returns an augmented response with the total number of matched items of the query. This `total_matches` value is critical to determine if we are trying to do a query-based selection or just selecting individual items when requesting a bulk operation from the client.
   
    ![customMediaType](https://user-images.githubusercontent.com/46503462/161790645-e9b261de-2e59-4d51-af93-0aa940fb9a70.gif)


  - Relaxed the parsing of boolean values in query parameters (https://github.com/galaxyproject/galaxy/pull/13555/commits/1397aa06449e6b722b93a7967c8b08e13f320e79) to accept also JSON style booleans ("true", "false") and not exclusively python style booleans ("True", "False")
---

# Frontend *(WIP)*
- [x] Update the history content selection system to support *query selection* in addition to just specific items.
- [x] Run `hide`, `unhide`, `delete`, `undelete` and `purge` operations in bulk over a search query or specific items selection.
- [x] Adapt history wide operations to use bulk operations endpoint
- [ ] Add error handling
- [x] Add client tests 

---

# Open questions
- [x] How do we handle collection building based on query selection?
  - *Discussed in chat*: We will enable collection building only for specific item selection (query selection not supported)

---

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Add basic schema models for history bulk operations

davelopez committed:
Add basic bulk operation action to HistoryContentsService

Currently implemented only for `hide` and `unhide`

davelopez committed:
Add API endpoint for history contents bulk operation

davelopez committed:
Add explicit item set support to bulk operations

When the payload contains specific history items, the bulk operation will be applied only to those and other dynamic filters will be ignored.

davelopez committed:
Add API test for bulk operation

Test explicit items selection in the payload

davelopez committed:
Refactor `FilterQueryParams` model

Extract pagination and value filtering fields to their own models to be reused.

davelopez committed:
Ignore pagination and order in bulk operation requests

All items in the dynamic query must be processed so there is no point in having `limit` and `offset`, also `order` is not relevant.

davelopez committed:
Add API test for dynamic query in bulk operations

davelopez committed:
Remove collection building options from bulk operations

The collection building takes place in the client, the API will deal only with operations on individual items.

davelopez committed:
Add `HistoryItemOperator`

To encapsulate operation definitions for history items

davelopez committed:
Use correct manager for collections in `HistoryItemOperator`

davelopez committed:
Add API test for all supported bulk operations

davelopez committed:
Add special `count` view to index endpoint

Using the `count` view will make the enpoint return 204 and provide the `total_matches` of the search query in the response headers

davelopez committed:
Add API test for `count` view in index endpoint

mvdbeek reviewed (commented):
This looks very nice. I wouldn't say my comments (exchange the view key for a different media-type, don't use q/qv) in the backend channel need to be addressed in this PR, this all looks very nice and clean

davelopez committed:
Include items in response when querying for matches

This will allow to get the items in the query and the total matches in a single requests so it's easier to keep them in sync.

davelopez committed:
Add support for query selection on `SelectedItems` component

davelopez committed:
Add `urlDataWithHeaders` to URL utils

davelopez committed:
Add total matches to HistoryItemsStore

davelopez committed:
Add `HistorySelectionStatus` to `HistoryPanel`

davelopez committed:
Run bulk operation on selection

Including query selection

davelopez committed:
Update history contents index API endpoint docs

davelopez committed:
Convert to bulk operation using query the history wide operations

davelopez committed:
Hide collection build options on query selection

For now building collections from selected items is only allowed for specific items, not query selection

davelopez committed:
Refactor bulk operations to use new filtering system

davelopez committed:
Cleanup unused code

guerler commented on client/src/components/History/CurrentHistory/HistoryOperations.vue:
```diff
@@ -88,7 +88,7 @@
                         <span v-localize>Delete All Hidden Content</span>
                     </b-dropdown-item>
                     <b-dropdown-item v-b-modal:purge-all-deleted-content>
-                        <span v-localize>Purge All Hidden Content</span>
```
I thought that this would purge all hidden datasets, not all deleted datasets?

davelopez replied:
I think this message was probably a copy&paste error from the one above, the actual defined function and the `v-b-modal` is `purge-all-deleted-content`.
Also, *deleting all hidden content* seems like an unsafe operation since usually, when you create a collection, the source datasets will be hidden and you may destroy them without being aware.

guerler commented on client/src/components/History/model/queries.js:
```diff
@@ -8,6 +8,7 @@
 import axios from "axios";
 import moment from "moment";
 import { prependPath } from "utils/redirect";
+import { getQueryDictFromFilters } from "../../../store/historyStore/historyItemsFiltering";
```
Please use an explicit path here.

guerler commented on client/src/components/History/model/queries.js:
```diff
@@ -272,23 +254,26 @@ export async function updateContentFields(content, newFields = {}) {
     const { id } = history;
-    const url = `/histories/${id}/contents`;
-    const payload = Object.assign({}, fields, { items });
+    const filterQuery = buildLegacyQueryStringFrom(filters);
```
I would not call it legacy, the word legacy is a little bit overused within Galaxy. Suggesting to call it just `buildQueryStringFrom`.

guerler commented on client/src/components/providers/storeProviders.js:
```diff
@@ -149,19 +150,23 @@ export const StoreProvider = (storeAction, storeGetter) => {
             result() {
                 return this[storeGetter](this.attributes);
             },
+            totalMatches() {
```
We should rename `queryStatsGetter` and `totalMatches`, since this is a generic store provider, accordingly the naming should probably also more generic. Maybe `storeGetterCount` and just `count` as parsed slot data parameter to indicate that they belong together?

davelopez replied:
I agree to make the naming more generic, and I have changed them to `storeCountGetter` and `count` respectively for now.
But I think just `count` may be confused with *the number of items returned in payload* which is not this value, this value represents *the total number of items that match the current query parameters without any pagination or limit*.
I'm happy to rename it to something more accurate, I always consider naming one of the hardest parts :laughing: 

Any better ideas?

guerler replied:
That looks fine to me. Thanks.

guerler commented on client/src/components/History/CurrentHistory/HistoryPanel.vue:
```diff
@@ -26,23 +29,35 @@
-                            @select-all="selectItems(payload)"
+                            @reset-selection="resetSelection"
                             @collapse-all="collapseAll" />
+                        <transition name="shutterfade">
```
I suggest to try it without the fade here. Right now it disappears in stages and there is a delay until the list of items is re-rendered.

guerler commented on client/src/components/History/CurrentHistory/HistoryOperations.vue:
```diff
@@ -209,12 +213,20 @@ export default {
-            await fn(this.history, type_ids);
-            this.$emit("hide-selection", items);
+            const items = this.getExplicitlySelectedItems();
+            const filters = getFilters(this.filterText);
```
We should probably use `getFilterDict` here instead, then there is no need for additional processing in the bulk query module and `getQueryDictFromFilters` can be removed again.

davelopez committed:
Remove shutterfade animation

davelopez committed:
Rename store getter and state for total matches

davelopez committed:
Allow JSON boolean when parsing query filters

This will avoid having to explicitly transform booleans in the client when building the query

davelopez committed:
Use `getQueryDict` instead of `getFilters` in `runOnSelection`

davelopez committed:
Redesign `HistorySelectionStatus` template

And simplify logic a bit

davelopez committed:
Mark as selected items that match the current query

When the query mode is enabled, all items that match the query will appear as selected as they load

davelopez committed:
Remove redundant prop from `HistorySelectionStatus`

davelopez committed:
Add tests for HistorySelectionStatus component

mvdbeek commented on lib/galaxy/webapps/galaxy/api/history_contents.py:
```diff
@@ -306,6 +311,20 @@ class FastAPIHistoryContents:
+                    },
+                },
+            },
+        },
```
Can you use a custom media type instead of `v=dev&view=count` ? We have to stop propagating these legacy behaviors.

davelopez replied:
Sure I can try!
Is setting the request content type to something like `header["content-type"] = "application/json-count"` and checking the header in the endpoint what you had in mind? Or I'm misunderstanding?

To be honest, it feels a bit weird to me messing with the content type, should we just add a different optional header or just add a new optional parameter to the endpoint?

By the way, what about the naming convention for headers? should we rename `total_matches` to `total-matches` or doesn't matter?

davelopez replied:
Thank you Marius for pointing me in the right direction. When you have a moment, please have a look at c26f6c9f0f276ae01941ebf50737a7283dd5d293 and let me know if there is something you would change.

mvdbeek commented on client/src/store/historyStore/historyItemsStore.js:
```diff
@@ -47,9 +51,13 @@ const actions = {
     fetchHistoryItems: async ({ commit, dispatch }, { historyId, offset, filterText }) => {
         dispatch("startHistoryChangedItems", { historyId: historyId });
         const queryString = getQueryString(filterText);
+        const queryStats = "&view=count";
```
This could have a significant impact on performance. Have you tested this on large histories ? Could this ever run on non-orm filters, where the ORM has to fetch every item from the database ?

mvdbeek replied:
> Could this ever run on non-orm filters, where the ORM has to fetch every item from the database ?

It looks like we'd discard any function filter in `contents_count` level, which is good for performance. genome_build, data_type and annotation are potentially valid filters though that are not being applied, so that could lead to wrong counts here. Not a big deal since I think we don't expose these in the UI, but I really think we need to pull out the filter parsing and fail the request if these are included.

davelopez replied:
I've tried just with 1000 datasets in the history, probably not large enough...

> Could this ever run on non-orm filters, where the ORM has to fetch every item from the database ?

To be honest, I was blindly trusting [content_counts](https://github.com/galaxyproject/galaxy/blob/7c442d63dfb91c748ee1ae3c0843fdc9a1543012/lib/galaxy/managers/history_contents.py#L120) to be the most optimised and tested way of counting all matching elements, since it was also used in https://github.com/galaxyproject/galaxy/pull/11436/files#diff-d7789a8b5b8606bd040d59409203acc178c14fe4ec4d6b38e8955d87dc4ecea1R666 for that purpose.

If this is not the way to go to know how many items match the current query... then I could really use your optimizing skills to do it right :pleading_face: :pray: :innocent:

davelopez replied:
> I really think we need to pull out the filter parsing and fail the request if these are included

Oh! didn't get your update while I was typing. Sure! I can try to make the request fail when those filters are present (just in the particular case of the count being requested).

mvdbeek replied:
> If this is not the way to go to know how many items match the current query... then I could really use your optimizing skills to do it right

puh, I don't know if it can be optimized further. I would rather think we should aggressively limit how often we have to run the query. Let's just go with it for now, we should know soon enough if this will be a problem.

davelopez replied:
Done in [c6bbbb1](https://github.com/galaxyproject/galaxy/pull/13555/commits/c6bbbb13ee6aba47cee7251926d528de50132380) :)

davelopez committed:
Replace custom `count` view with custom media type

The history contents `index` endpoint will now accept a custom media type (`application/vnd.galaxy.history.contents.stats+json`) that will return a different response model containing additional item counting stats (`total_matches` for now) in addition to the regular items in the response

davelopez committed:
Raise Bad Request when requesting stats with non-orm filters

For details see https://github.com/galaxyproject/galaxy/pull/13555#discussion_r842646914

davelopez committed:
Remove unused functions from History queries

davelopez committed:
Make destructive purge operation slightly more obvious

bgruening reviewed (approved):
Very nice work @davelopez! Tested it locally and it works nicely for me.

Thanks especially for the bunch of tests!
